[16.0][MIG] product_multi_company: Migration to version 16.0

pilarvargas-tecnativa opened the pull request:
Superseed #459 because the PR is very old and the creator stopped taking care of it, this PR recovers the changes made in its migration but following all the steps of a conventional migration and finalising the migration.

cc @Tecnativa TT48972

@pedrobaeza @carolinafernandez-tecnativa please review

pedrobaeza committed:
product_multi_company

==========================================
Product permissions for discrete companies
==========================================

This modules allows to select in which of the companies you want to use each
of the products.

Installation
============

This module uses the post and uninstall hooks for updating default product
template security rule. This only means that updating the module will not
restore the security rule this module changes. Only a complete removal and
reinstallation will serve.

Usage
=====

On the product form view, go to the "Information" tab, and put the companies
in which you want to use that product. If none is selected, the product will
be visible in all of them. The default value is the current one.

pedrobaeza committed:
product_multi_company: Migration to 9.0

oca-transbot committed:
OCA Transbot updated translations from Transifex

lasley committed:
[MIG] product_multi_company: Upgrade to v10
* Rename manifest
* Change openerp references to odoo
* Bump version
* Add pragma no cover
* Edit security of product employee to allow writes (in tests)
* Fix permissions in tests
* Fix domain & add test
* Implement base_multi_company on product_multi_company
* Add related cols for product variant

oca-transbot committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[MIG] *_multi_company: OpenUpgrade migration scripts

oca-transbot committed:
OCA Transbot updated translations from Transifex

cubells committed:
[MIG] product_multi_company: Migration to 11.0

pedrobaeza committed:
[MIG] product_multi_company: Remove migration scripts

florian-dacosta committed:
Migrate product_multi_company to version 12

florian-dacosta committed:
Fix product_multi_company description

JoanSForgeFlow committed:
[IMP] product_multi_company: black, isort, prettier

JoanSForgeFlow committed:
[MIG] product_multi_company: Migration to 13.0

AaronHForgeFlow committed:
[MIG]product_multi_company: Migration to 13.0 (continue)

Commit:
[FIX] product_multi_company: integrate tests with product_variant_configurator

When this module is integrated with `product_variant_configurator`, tests fail because the latter makes some black magic on the `name` field.

I switch the tests to rely on the `description_sale` field instead, which isn't touched by that module, and is also a field delegated from the related `product.template`.

BTW it doesn't make sense to modify provided permissions on a test that checks access errors, so instead I switched them to use users in a group that should have access in production.

@Tecnativa TT27288

OCA-git-bot committed:
product_multi_company 13.0.1.0.1

anasuarez1 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (4 of 4 strings)

Translation: multi-company-13.0/multi-company-13.0-product_multi_company
Translate-URL: https://translation.odoo-community.org/projects/multi-company-13-0/multi-company-13-0-product_multi_company/es/

rousseldenis committed:
[IMP] product_multi_company: black, isort, prettier

rousseldenis committed:
[14.0][MIG] product_multi_company

rousseldenis committed:
[14.0][IMP] product_multi_company: Add tests common class

In order to extend testing, a common class is necessary to avoid code
duplication.

rousseldenis committed:
[14.0][IMP] product_multi_company: Create products not shared in correct context

As some default values can come from current company (e.g.: main one), change context
when creating.

rousseldenis committed:
[14.0][FIX] product_multi_company: Add no_company_ids field

rousseldenis committed:
[14.0][FIX] product_multi_company: Remove migration script

rousseldenis committed:
[40.0][IMP] product_multi_company: Add methods helpers for tests

JasminSForgeFlow committed:
[MIG] product_multi_company: Migration to 15.0

prabakaran04 committed:
[MIG] product_multi_company: Migration to 16.0

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: multi-company-15.0/multi-company-15.0-product_multi_company
Translate-URL: https://translation.odoo-community.org/projects/multi-company-15-0/multi-company-15-0-product_multi_company/

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (7 of 7 strings)

Translation: multi-company-15.0/multi-company-15.0-product_multi_company
Translate-URL: https://translation.odoo-community.org/projects/multi-company-15-0/multi-company-15-0-product_multi_company/es/

pilarvargas-tecnativa committed:
[IMP] product_multi_company: pre-commit stuff

pilarvargas-tecnativa committed:
[MIG] product_multi_company: Continue migration to version 16.0

TT48972

pedrobaeza commented:
/ocabot migration product_multi_company

carolinafernandez-tecnativa commented on product_multi_company/readme/USAGE.rst:
```diff
@@ -0,0 +1,3 @@
+On the product form view, go to the "Information" tab, and put the companies
```
tab "Information" renamed for "General Information"

carolinafernandez-tecnativa commented on product_multi_company/models/product_category.py:
```diff
@@ -0,0 +1,29 @@
+        "route_id",
+        "Routes",
+        domain=lambda self: [
+            "|",
+            ("company_id", "=", False),
+            ("company_id", "in", self.env.companies.ids),
```
better use [('company_id', 'in', self.env.companies.ids + [False])]

pilarvargas-tecnativa replied:
thanks

carolinafernandez-tecnativa commented on product_multi_company/models/product_category.py:
```diff
@@ -0,0 +1,29 @@
+
+    total_route_ids = fields.Many2many(
+        domain=lambda self: [
+            "|",
+            ("company_id", "=", False),
+            ("company_id", "in", self.env.companies.ids),
```
better use [('company_id', 'in', self.env.companies.ids + [False])]

carolinafernandez-tecnativa commented on product_multi_company/hooks.py:
```diff
@@ -0,0 +1,34 @@
+    rule.write(
+        {
+            "domain_force": (
+                "['|', ('company_ids', 'in', company_ids),"
+                "('company_ids', '=', False)]"
+            ),
```
same here

carolinafernandez-tecnativa commented:
Functional review LGTM :)

Please rebase and check for CI.
Do not forget to squash administrative commits. Thanks!

amkarthik commented:
@carolinafernandez-tecnativa 

We are getting this issue in this PR:
https://github.com/OCA/multi-company/issues/252

Please check the attached video below and that is recorded in PR runboat.

https://github.com/OCA/multi-company/assets/12694917/024b9513-779c-4266-9e83-8a5e8d6591a8

Thanks

pilarvargas-tecnativa commented:
> We are getting this issue in this PR: #252

It may have been something transient in the runbot, since duplicating the user as ‘User B’ and repeating the same operation does not cause this problem.

pedrobaeza commented on product_multi_company/models/product.py:
```diff
@@ -1,19 +0,0 @@
-class ProductProduct(models.Model):
-    _inherit = "product.product"
-
-    company_ids = fields.Many2many(
```
Why removing this?

pilarvargas-tecnativa replied:
It was not in v15, it is not necessary

pilarvargas-tecnativa replied:
The fact is that by extending the abstract model in product.template, product.product acquires that field.

amkarthik replied:
@pilarvargas-tecnativa 
From what I know, the above concept is the reason for the user A and user B issue.
As product.product doesn't hold the company_ids directly in it, if someone tries to access the product.product then it tries to compute it from the product.template.
This causing the above issue as the user doesn't have access to all the companies that is set on the product.template.

amkarthik commented:
> > We are getting this issue in this PR: #252
> 
> It may have been something transient in the runbot, since duplicating the user as ‘User B’ and repeating the same operation does not cause this problem.

Hi I tested again with the User B. (changed the password of userb to userb).
Same issue happens to me.
Please check.


https://github.com/OCA/multi-company/assets/12694917/7c930b43-2e2a-4a04-8f16-6db26143eda7

pilarvargas-tecnativa commented:
> Hi I tested again with the User B. (changed the password of userb to userb). Same issue happens to me. Please check.

I have retested the test locally, checking in shell the data contained in the bd if it contains the company_ids field. 
![image](https://github.com/OCA/multi-company/assets/118818446/c5085144-eed9-4770-8b71-9ae870ae2e9b)
![image](https://github.com/OCA/multi-company/assets/118818446/a26ca6eb-1607-47e8-8255-2c5ec54b5490)

I have replicated the error but it only happens in the case of new companies created, it does not happen with the demo companies already defined in odoo. I am still investigating the error.

amkarthik commented:
> > Hi I tested again with the User B. (changed the password of userb to userb). Same issue happens to me. Please check.
> 
> I have retested the test locally, checking in shell the data contained in the bd if it contains the company_ids field. ![image](https://private-user-images.githubusercontent.com/118818446/333520796-c5085144-eed9-4770-8b71-9ae870ae2e9b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY1NDM1MjAsIm5iZiI6MTcxNjU0MzIyMCwicGF0aCI6Ii8xMTg4MTg0NDYvMzMzNTIwNzk2LWM1MDg1MTQ0LWVlZDktNDc3MC04YjcxLTlhZTg3MGFlMmU5Yi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUyNFQwOTMzNDBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00OWU3YzMyNTAzYzhmYWM4OGEzZWUzN2JlNzgxZWI2MjlmNjNlZTVkMWI2OTk4N2U2MDJjMzJjOTQ5YjFmM2U0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.mDHJ3koBZ_J6cwbYZ-AEnJUPs5M8D2oVZCqkJoXxWCM) ![image](https://private-user-images.githubusercontent.com/118818446/333521033-a26ca6eb-1607-47e8-8255-2c5ec54b5490.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY1NDM1MjAsIm5iZiI6MTcxNjU0MzIyMCwicGF0aCI6Ii8xMTg4MTg0NDYvMzMzNTIxMDMzLWEyNmNhNmViLTE2MDctNDdlOC04MjU1LTJjNWVjNTRiNTQ5MC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUyNFQwOTMzNDBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iZDk4NDE3ZTQ3MGVmMTU0NmE0NzBiMTA1NWQyOTdlNDY2NDgxZTIzMzc1NmI2NGI3M2FjZDEyZjgwMTYzNjcwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.4eAMyDRvyoxOXD700X_gydJWP00_OF-aJSZM3aoG5PQ)
> 
> I have replicated the error but it only happens in the case of new companies created, it does not happen with the demo companies already defined in odoo. I am still investigating the error.

Hi,
Can you please confirm the user that you are using in the shell?

pilarvargas-tecnativa commented:
@amkarthik Actually, the access rule to the ‘product.product’ model was not being applied and it is necessary to do it as it was done in v13 and v14. This same error already occurs in v15 where this part of the code has been removed. This error should no longer appear and I have added the corresponding unit tests so that it does not happen again. Thanks

carolinafernandez-tecnativa commented:
Please check CI, still having problems with multi-company rules, thanks!

amkarthik commented:
> @amkarthik Actually, the access rule to the ‘product.product’ model was not being applied and it is necessary to do it as it was done in v13 and v14. This same error already occurs in v15 where this part of the code has been removed. This error should no longer appear and I have added the corresponding unit tests so that it does not happen again. Thanks

Thanks

amkarthik commented on product_multi_company/models/product_product.py:
@pilarvargas-tecnativa 
Why we are going back to using create and write method as in v13.
Why don't we consider creating related field instead?
    company_ids = fields.Many2many(
        comodel_name="res.company",
        column1="product_id",
        column2="company_id",
        relation="product_product_company_rel",
        related="product_tmpl_id.company_ids",
        compute_sudo=True,
        readonly=False,
        store=True,
    )

This is in the PR #459

pilarvargas-tecnativa replied:
Having debugged all the code after all the conflicts with tests of other modules that uncovered the errors that occurred in this one, it is certain that this option is the most appropriate. The tests to check that the variants have their access permissions correctly set are also maintained. Thanks

carolinafernandez-tecnativa commented:
With these changes, when creating new companies and accessing with created users to product variant the error seems to disappear.
I also have checked accessing in product template and now it seems to work fine, Thanks !

pedrobaeza commented on product_multi_company/__manifest__.py:
```diff
@@ -7,10 +7,12 @@
     "depends": ["base_multi_company", "product", "stock"],
-    "data": ["views/product_template_view.xml"],
+    "data": [
+        "views/product_template_view.xml",
```
This change is not needed. Please undo in the corresponding commit.

pedrobaeza commented on product_multi_company/__manifest__.py:
```diff
@@ -0,0 +1,18 @@
+    "category": "Product Management",
+    "version": "16.0.1.0.0",
+    "license": "AGPL-3",
+    "depends": ["base_multi_company", "product", "stock"],
```
The dependency on stock must be removed.

pedrobaeza commented on product_multi_company/models/product_category.py:
```diff
@@ -8,22 +8,14 @@ class ProductCategory(models.Model):
 
     route_ids = fields.Many2many(
-        "stock.location.route",
+        "stock.route",
```
Remove these overrides. That should go if ever done on a module `product_multi_company_stock`.

pilarvargas-tecnativa replied:
All done! thanks

pilarvargas-tecnativa committed:
[IMP-FIX] product_multi_company: Remove dependency to stock

Remove dependency to stock and its overwritten fields, this logic should
go  if ever done in a separate module named: product_multi_company_stock

TT48972

pedrobaeza reviewed (approved):
/ocabot merge nobump

cyrilmanuel commented:
Hello @pedrobaeza, could you merge this PR ?
